packages/devtools_app/lib/src/standalone_ui/api/impl/vs_code_api.dart

```diff
@@ -0,0 +1,149 @@
+// Copyright 2023 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:json_rpc_2/json_rpc_2.dart' as json_rpc_2;
+import 'package:meta/meta.dart';
+
+import '../vs_code_api.dart';
+import 'dart_tooling_api.dart';
+
+final class VsCodeApiImpl extends ToolApiImpl implements VsCodeApi {
+  VsCodeApiImpl._(super.rpc, Map<String, Object?> capabilities) {
+    this.capabilities = VsCodeCapabilitiesImpl(capabilities);
+    devicesChanged =
+        events('devicesChanged').map(VsCodeDevicesEventImpl.fromJson);
+  }
+
+  static Future<VsCodeApi?> tryConnect(json_rpc_2.Peer rpc) async {
+    final capabilities = await ToolApiImpl.tryGetCapabilities(rpc, _apiName);
+    return capabilities != null ? VsCodeApiImpl._(rpc, capabilities) : null;
+  }
+
+  static const _apiName = 'vsCode';
+
+  @override
+  Future<void> initialize() => sendRequest('initialize');
+
+  @override
+  @protected
+  String get apiName => _apiName;
+
+  @override
+  late final Stream<VsCodeDevicesEvent> devicesChanged;
+
+  @override
+  late final VsCodeCapabilities capabilities;
+
+  @override
+  Future<Object?> executeCommand(String command, [List<Object?>? arguments]) {
+    return sendRequest(
+      'executeCommand',
+      {'command': command, 'arguments': arguments},
+    );
+  }
+
+  @override
+  Future<bool> selectDevice(String id) {
+    return sendRequest(
+      'selectDevice',
+      {'id': id},
+    );
+  }
+}
+
+class VsCodeDeviceImpl implements VsCodeDevice {
```

Comment:
The goal was to make the API much clearer so it can more easily be mirrored/compared to VS Code. Without interfaces it might also be quite easy to make an accidental breaking change to the interface (and the mock environment) without VS Code.

I added some more comments to the interface classes to make this clearer, although I'm not married to the idea - if you'd prefer not to have them (and aren't concerned about accidental breakage), I'm happy to remove :-)